examples/channel_inversion/hessian_evaluation.py

```diff
@@ -0,0 +1,112 @@
+import firedrake
+import ufl
+
+
+class HessianEstimator:
+    def __init__(self, f, mesh=None, target_spaces=None):
+        r"""
+        Recover the gradient and Hessian of a scalar field using a
+        double :math:`L^2` projection.
+        :arg f: the scalar field whose derivatives we seek to recover
+        :kwarg mesh: the underlying mesh
+        :kwarg target_spaces: the :class:`VectorFunctionSpace` and
+            :class:`TensorFunctionSpace` the recovered gradient and
+            Hessian should live in
+        """
+        mesh = mesh or f.function_space().mesh()
+        if target_spaces is None:
+            P1_vec = firedrake.VectorFunctionSpace(mesh, "CG", 1)
+            P1_ten = firedrake.TensorFunctionSpace(mesh, "CG", 1)
+        else:
+            P1_vec, P1_ten = target_spaces
+        W = P1_vec*P1_ten
+        g, H = firedrake.TrialFunctions(W)
+        phi, tau = firedrake.TestFunctions(W)
+        l2_projection = firedrake.Function(W)
+        n = ufl.FacetNormal(mesh)
+
+        self.gradient, self.hessian = l2_projection.split()
+
+        # The formulation is chosen such that f does not need to have any
+        # finite element derivatives
+        a = ufl.inner(tau, H)*ufl.dx \
+            + ufl.inner(ufl.div(tau), g)*ufl.dx \
+            - ufl.dot(g, ufl.dot(tau, n))*ufl.ds
+        a += ufl.inner(phi, g)*ufl.dx
+        L = f*ufl.dot(phi, n)*ufl.ds \
+            - f*ufl.div(phi)*ufl.dx
```

Comment:
I noticed how you added in flux terms for the weak gradient below. Makes sense in terms of generalisability, so I [added for Hessian](https://github.com/pyroteus/pyroteus/blob/88f72c6dd827604ba5842aa98ea9183236d71afd/pyroteus/recovery.py#L65-L72) too. Might be worth doing here (since many Thetis fields are DG)?